implementations/rust/ockam/ockam/src/pipe2/mod.rs

```diff
@@ -0,0 +1,159 @@
+//! replace it at some point in the future.
+
+mod hooks;
+mod receiver;
+mod sender;
+
+use crate::{Context, OckamMessage, SystemBuilder, SystemHandler};
+use ockam_core::{
+    compat::{boxed::Box, string::String, vec::Vec},
+    Address, Result, Route,
+};
+
+enum Mode {
+    /// In static mode this pipe will connect to a well-known peer, or
+    /// receive _one_ connection on a well-known address, or does
+    /// both at the same time
+    Static,
+    /// In dynamic mode this pipe connects to a peer via an
+    /// initialisation handshake, or listens for initialisation
+    /// handshakes
+    Dynamic,
+}
+
+/// A builder structure for pipes
+///
+/// The easiest way to create a pipe is via
+/// `builder.basic().static()`.
+pub struct PipeBuilder {
+    send_hooks: SystemBuilder<Context, OckamMessage>,
+    recv_hooks: SystemBuilder<Context, OckamMessage>,
+    /// The selected pipe initialisation mode
+    mode: Mode,
+    /// Peer information
+    peer: Option<Route>,
+    /// Receiver information
+    recv: Option<Address>,
+    /// This address is used to handle the termination point of the
+    /// worker system pipeline.
+    fin: Address,
+}
+
+impl PipeBuilder {
+    fn new(mode: Mode) -> Self {
+        Self {
+            send_hooks: SystemBuilder::new(),
+            recv_hooks: SystemBuilder::new(),
+            peer: None,
+            recv: None,
+            fin: Address::random(0),
+            mode,
+        }
+    }
+
+    /// Construct a fixed pipe to a specific well-known peer
+    pub fn fixed() -> Self {
+        Self::new(Mode::Static)
+    }
+
+    /// Construct a pipe using dynamic initialisation handshakes
+    pub fn dynamic() -> Self {
+        Self::new(Mode::Dynamic)
+    }
+
+    /// Set a connection peer, creating an outgoing pipe
+    ///
+    /// * In `fixed` mode this attempts to connect directly to a
+    /// receiver at the given peer route.
+    ///
+    /// * In `dynamic` mode this initiates a handshake with the
+    /// given peer.  This handshake then resolves to the final
+    /// receiver
+    pub fn connect<R: Into<Route>>(mut self, peer: R) -> Self {
+        self.peer = Some(peer.into());
+        self
+    }
+
+    /// Set a receiving address, creating a receiving pipe
+    ///
+    /// * In `fixed` mode this creates a pipe receiver which waits
+    /// for incoming messages from a sender.
+    ///
+    /// * In `dynamic` mode this spawns a handshake listener, which
+    /// will create pipe receivers dynamically for any incoming
+    /// initialisation request
+    pub fn receive<A: Into<Address>>(mut self, addr: A) -> Self {
+        self.recv = Some(addr.into());
+        self
+    }
+
+    /// Set this pipe to enforce the ordering of incoming messages
+    pub fn enforce_ordering(mut self) -> Self {
+        // A pipe can currently only have two types of hooks.
+        // "Delivery" and "Ordering".  Because we want "ordering" to
```

Comment:
Minor note, but this sounds like something to be done on `build`

Reply:
I was kinda on the fence about this. On one hand, yes, it feels like this is a bit too complex to be done in-line to these attribute functions. And moving this code to `build` will also simplify it.

On the other hand it does limit the behaviour that users can specify. But maybe that's okay. I opted for the slightly more flexible and complicated solution to start